### ISSUE TITLE
Infinite loop on single core, SCHED_FIFO, latched message, tcpNoDelay

### ISSUE THREAD

**Issue by TinyTheBrontosaurus** [ros/ros_comm#1429]
A ROS node will get stuck at 100% CPU if the following occur:

* Subscribe done with udp().tcpNoDelay(true)
* A Latched message waiting at subscribe time
* Process set at SCHED_FIFO
* Single core available

See minimal reproducible example here: https://github.com/TinyTheBrontosaurus/rosbug_latched_singlecore

Seen in Ubuntu 16.04 with ROS Kinetic

**Comment by cwecht** [comment]
I would strongly discourage using ``SCHED_FIFO`` with ROS nodes. The internals of ROS are using boost mutexes, which can not use priority inheritance, because boost does not provide this feature. This is required, because ``SCHED_FIFO`` processes can not be preempted by processes with the same priority. Boost locks might use spinlocks, which can lead to livelocks. 

If you want to use realtime scheduling, ``SCHED_RR`` should be the much better choice.

Basically the problem we are facing here arises from the combination of ``SCHED_FIFO`` and pinning the process to a single CPU. Either not using ``SCHED_FIFO``or pinning the processes to more than one CPU fixes the issue.

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
